### ISSUE TITLE
Loops with "as parallel" expression:

### ISSUE THREAD

**Issue by ghost** [dotnet/csharplang#2188]
I suggest to write parallel for loops (and other types of loops if possible) as this:

```
for(i = 0; i < files.Length; i++) as parallel
{ 
      FileInfo fi = new FileInfo(files[i]);
      long size = fi.Length;
      Interlocked.Add(ref totalSize, size);
} 
```
Which can easily converted by C# compiler to:

```
Parallel.For(0, files.Length,
      i => { 
                 FileInfo fi = new FileInfo(files[i]);
                 long size = fi.Length;
                 Interlocked.Add(ref totalSize, size);
              } );
```

It maybe a small difference, but the former is more familiar and easier to read.

**Comment by CyrusNajmabadi** [comment]
Yeah, if this were to be done, it would basically have to be:

```c#
parallel for( i = 0; i < files.Length; i++)
{ 
      FileInfo fi = new FileInfo(files[i]);
      long size = fi.Length;
      Interlocked.Add(ref totalSize, size);
} 
```

**Comment by HaloFour** [comment]
> It is just a syntax that can be interpreted into the original API code.  

Except that it can't.  You can't specify some arbitrary condition with `Parallel.For`.  You can't specify some arbitrary increment with `Parallel.For`.  What would the compiler interpret the following into?

```cs
bool flag = true
parallel for (int i = 0; i < files.Length || flag; i += 2) {
    flag = false;
}
```

That would have to be a compiler error because it can't be translated into `Parallel.For`.  But that's perfectly legal `for` syntax otherwise.

How do you specify the degree of parallelism?  A cancellation token?  What scheduler to use?  You can't, without more syntax.

> This will programmers to try parallelism, and enhance their slow codes by just adding a word or two.

Or make it a whole lot slower because throwing threads at a problem doesn't make it "faster".  Threads add a lot of complexity, including overhead and synchronization concerns.  Walking into parallelism haphazardly is a surefire way to make slower and buggier code.

**Comment by MohammadHamdyGhanem** [comment]
@HaloFour:
As you said:  Put these roles in docs and give a compiler error if they don't apply.
Put I want to \ask: why can't a loop wuth an increament be done in parralel? If this idea to be done, it will be a good thing to allow adding the increment in the API. I can say the same about a cancelation condition like (flag == true). But these things can be postponed to another time after making the basic syntax.

**Comment by HaloFour** [comment]
>  Put these roles in docs and give a compiler error if they don't apply.

What's the point of trying to make it look like a `for` loop if you're required by the compiler to use a single comparison and a single increment statement?  That completely defeats the idea of being able to "parallelize" your loops just by slapping a `parallel` before them.

>why can't a loop wuth an increament be done in parralel?

Because `Parallel.For` doesn't have an overload to do that.

> I can say the same about a cancelation condition like (flag == true).

No you can't, the API takes a cancellation token in a very specific way.  Any alternative syntax would need to offer all of the same permutations, and be able to handle future permutations.  `ParallelOptions` exists for a reason.

> But these things can be postponed to another time after making the basic syntax.

Which will never happen.  There's no need to expend a ton of effort trying to redesign what already exists perfectly fine as a library method call.

**Comment by yaakov-h** [comment]
> why can't a loop wuth an increament be done in parralel?

It can, but it's sufficiently complex that it has a specialized API and can't easily be generalized into a language feature.

**Comment by MohammadHamdyGhanem** [comment]
What about a new syntax for "parallel for" to fit the conditions:
```
parallel for (int i = 0 to files.Length -1)
{

}
```

??

**Comment by yaakov-h** [comment]
That would be the same as this, right?

```csharp
Parallel.For(0, files.Length - 1, i =>
{

});
```

Why add language syntax to make your code three characters _longer_?

**Comment by CyrusNajmabadi** [comment]
> What about a new syntax for "parallel for" to fit the conditions:

We can 'what about' to the end of time.  If you want this language feature, you have to provide *really compelling points* as to why this is a good idea.  Note: if you'd just like to discuss things and bounce ideas off of people, i'd recommend gitter.im/dotnet/csharplang.

**Comment by MkazemAkhgary** [comment]
not any code can be magically converted to parallel, most of the times in your loop order is important and result of each iteration depends on result previous iteration. 

note that GPUs can run simple instructions in parallel by design. but CPU is designed to be concurrent and not parallel. that's why making simple loops parallel slows down your program rather than speeding it up.

if your loop can run in parallel you must see if each iteration finishes quickly or not. if it's quick then you should divide your loop in medium or bigger batches and run them in parallel. you must still measure the performance.

see it's not as easy as your proposed syntax. (but it would be awesome in magic land)

**Comment by YairHalberstadt** [comment]
Closing as the OP no longer has a github account and this was fairly poorly received.
